service/entityresolution/integration/internal/chain_contract_tests.go

```diff
@@ -0,0 +1,255 @@
+package internal
+
+import (
+	"context"
+	"strings"
+	"testing"
+
+	"connectrpc.com/connect"
+	"github.com/opentdf/platform/protocol/go/entity"
+	entityresolutionV2 "github.com/opentdf/platform/protocol/go/entityresolution/v2"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// ChainContractTestSuite holds implementation-agnostic multi-entity chain validation tests
+type ChainContractTestSuite struct {
+	TestCases []ContractTestCase
+}
+
+// NewChainContractTestSuite creates a test suite focused on implementation-agnostic multi-entity chain validation
+func NewChainContractTestSuite() *ChainContractTestSuite {
+	return &ChainContractTestSuite{
+		TestCases: []ContractTestCase{
+			{
+				Name:        "CreateMultiEntityChainFromSingleToken",
+				Description: "Should create entity chain with multiple entities and proper categorization",
+				Input: ContractInput{
+					Entities: []*entity.Entity{},
+					Tokens: []*entity.Token{
+						CreateTestToken("chain-token-1", "test-client-1", "alice", "alice@opentdf.test"),
+					},
+				},
+				Expected: ContractExpected{
+					EntityCount:      0,
+					ShouldError:      false,
+					ErrorCode:        0,
+					EntityValidation: []EntityValidationRule{},
+					ChainValidation: []EntityChainValidationRule{
+						{
+							EphemeralID:              "chain-token-1",
+							EntityCount:              2, // Both Keycloak and Multi-Strategy create 2 entities per token
+							EntityTypes:              []string{}, // Implementation-agnostic: don't specify entity types
+							EntityCategories:         []string{"CATEGORY_ENVIRONMENT", "CATEGORY_SUBJECT"}, // Both must create these categories
+							RequireConsistentOrdering: false, // Allow flexible ordering between implementations
+						},
+					},
+				},
+			},
+			{
+				Name:        "CreateMultiEntityChainsFromMultipleTokens",
+				Description: "Should create multiple entity chains with consistent multi-entity behavior",
+				Input: ContractInput{
+					Entities: []*entity.Entity{},
+					Tokens: []*entity.Token{
+						CreateTestToken("chain-token-1", "test-client-1", "alice", "alice@opentdf.test"),
+						CreateTestToken("chain-token-2", "test-client-2", "bob", "bob@opentdf.test"),
```

Comment:
so the tokens created here would be the same i think except for the ephemeral id, but the info inside would be the same?

Reply:
This is fixed